lib/signalfx/signalflow/channel.rb

```diff
@@ -1,139 +1,88 @@
 # The main method for it is {#each_message}, which is how you get messages from
-# the channel.
+# the channel.  There can only be one user of a channel and they are NOT
+# thread-safe.
 #
 # Channels are for one-time use only.  Once a channel is detached from (either
 # manually or due to the end of a computation) previous messages will be
 # iterable but nothing new will show up.
 class Channel
   attr_accessor :name
-  attr_accessor :computation
   attr_accessor :detached
 
   def initialize(name, detach_cb)
     @lock = Mutex.new
     @detach_lock = Mutex.new
-    @cb_lock = Mutex.new
     @detached = false
     @name = name
     @detach_from_transport = detach_cb
-    @cbs = []
-    # Cache every message that comes in on this channel so that we can replay
-    # them when iterating messages and avoid races.
-    @messages = []
-  end
-
-  def computation=(comp)
-    @computation = comp
+    @messages = Queue.new
   end
 
-  # Call a block with each message asynchronously.  Returns immediately.
+  # Waits for and returns the next message in the channel.
   #
-  # @yield [msg, detach] Called when a message arrives that is relevant to the
-  #   channel's computation.  The `detach` param will be set to a function that can
-  #   be called by the block to detach from the computation.
-  def each_message_async(&block)
-    raise 'each_message_async() requires block' unless block
-
-    @cb_lock.synchronize do
-      replay_existing_messages(&block)
-      return if @detached
-
-      @cbs << block
-    end
-    return
-  end
-
-  # Call the given block with each message in the channel as they arrive.  This
-  # method will not return until the channel is detached from (either manually
-  # or due to the computation ending).
+  # @param timeout_seconds [Float] Number of seconds to wait for a message.
   #
-  # @yield [msg, detach] Called when a message arrives that is relevant to the
-  #   channel's computation.  The `detach` param will be set to a function that can
-  #   be called by the block to detach from the computation.
-  def each_message(&block)
-    raise 'each_message() requires block' unless block
-
-    wait_lock = Mutex.new
-    cv = ConditionVariable.new
-
-    @cb_lock.synchronize do
-      replay_existing_messages(&block)
-
-      return if @detached
-
-      @cbs << ->(msg, detach) do
-        send_message_to_block(msg, &block)
-        if @detached
-          wait_lock.synchronize do
-            cv.signal
+  # @return [Hash] The next message received by this channel.  A return value
+  # of `nil` indicates that the channel has detected it is done and will not be
+  # receiving any more useful messages.
+  #
+  # @raise [ChannelTimeout] If the timeout is exceeded with no messages
+  def pop(timeout_seconds=nil)
+    raise "Channel #{@name} is detached" if @detached
+
+    msg = nil
+    # Unfortunate Ruby queues don't support waiting with a timeout so hack it
+    # up ourselves so we don't end up blocking forever by accident.
```

Comment:
I don't think this will work well, because you'll incur a forced 100ms sleep if you happen to attempt a pop when nothing is in the queue, even if something arrives 1ms after. You might want to use a custom queue instead, like the implementation shown at https://spin.atomicobject.com/2017/06/28/queue-pop-with-timeout-fixed/

Reply:
Yeah I saw that when I was first writing it but went with timeout for simplicity originally, but I swiped that implementation in the link and am using that now so it should be more efficient.